examples/universal_request_tracing/README.md

```diff
@@ -0,0 +1,355 @@
+        # request_id is guaranteed to be non-None by the decorator
+        self.log("info", "Starting token decoding")
+
+        async for token_result in self._decode_tokens(hidden_states):
+            yield token_result
+```
+
+## Testing
+
+```python
+import asyncio
+import aiohttp
+
+async def test_universal_tracing():
+    headers = {
+        "Content-Type": "application/json",
+        "X-Request-Id": "test-universal-123"
+    }
+
+    payload = {
+        "model": "meta-llama/Llama-3.2-3B-Instruct",
+        "messages": [{"role": "user", "content": "Hello!"}],
+        "max_tokens": 50
+    }
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
+            "http://localhost:8080/v1/chat/completions",
+            headers=headers,
+            json=payload
+        ) as response:
+            # Verify X-Request-Id is correctly echoed back
+            assert response.headers.get("X-Request-Id") == "test-universal-123"
+            print("✅ Universal request tracing works!")
+
+if __name__ == "__main__":
+    asyncio.run(test_universal_tracing())
+```
+
+## Advantages
+
+### 1. Zero Configuration
+- Just add decorator or inherit Mixin class
+- No manual handling of HTTP headers or responses
+- Automatic request ID propagation
+
+### 2. Consistency
+- All components use the same request ID format
+- Unified logging format and tracing approach
+- Compatible with vLLM and OpenAI standards
+
+### 3. Flexibility
+- Supports class decorators, method decorators, and Mixin classes
+- Can get current request ID anywhere
+- Supports manual request ID passing
+
+### 4. Performance
+- Uses thread-local storage, no global locks
+- Minimal performance overhead
+- Automatic context cleanup
+
+## Migration Guide
+
+### Migrating from Manual Implementation
+
+**Before (Manual Implementation):**
+```python
+def extract_or_generate_request_id(request: Request) -> str:
+    request_id = request.headers.get("x-request-id")
+    if request_id is None:
+        request_id = str(uuid.uuid4())
+    return request_id
+
+@endpoint(is_api=True)
+async def chat_completions(self, request: Request, chat_request: ChatCompletionRequest):
+    request_id = extract_or_generate_request_id(request)
+
+    async def content_generator():
+        async for response in self.processor.chat_completions(chat_request, request_id):
+            yield response
+
+    response = StreamingResponse(content_generator())
+    response.headers["X-Request-Id"] = request_id
+    return response
+```
+
+**After (Using Universal Support):**
+```python
+@auto_trace_endpoints
+class Frontend:
+    @endpoint(is_api=True)
+    async def chat_completions(self, request: Request, chat_request: ChatCompletionRequest):
+        # Everything is handled automatically!
+        async for response in self.processor.chat_completions(chat_request):
+            yield response
+```
+
+## Important Notes
```

Comment:
probably best formatted as

```
> [!Important]
> 1. **FastAPI Request Object**: 
>     Frontend endpoints need to include `Request` parameter to extract HTTP headers
> 2. **Optional Parameters**: 
>     Processor methods should accept `Optional[str] request_id` parameter
> 3. **Thread Safety**: 
>     Uses thread-local storage, safe in async environments
> 4. **Context Propagation**: 
>     Router and Worker components automatically get request ID through Dynamo's Context system
> 
> This universal approach greatly simplifies X-Request-Id implementation, allowing developers to focus on business logic rather than infrastructure code.
```